doc/ledger-consensus-api.md

```diff
@@ -0,0 +1,146 @@
+    | LedgerRewards !EpochNo !Rewards
+    | LedgerStakeDist !EpochNo !StakeDist
+    | LedgerBlock !AnnotatedBlock
+  ````
+
+  Rewards and stake distribution could be calculated incrementally by the ledger and the partial
+  results could be passed to `db-sync` as they become available.
+
+With regards to rewards, there are currently two sources; rewards from minting blocks and
+instantaneous rewards. Currently these are two separate data structures, but are only distributed
+at epoch boundaries. In this new API, it would make sense to merge these two, with each separate
+reward being annotated by its source (eg pool owner, pool member, instantaneous treasury and
+instantaneous reserve payment).
+
+There would then be something like the `LocalChainSync` protocol that passes `LedgerEvent`
+objects over the connection rather than the current blockchain version of the blocks.
+
+It is important to note that the `cardano-wallet` *also* maintains its own copy of ledger state and
+*also* uses the local state query mechanism. Providing this new API would mean that both `db-sync`
+and `cardano-wallet` can remove large chunks of relatively complex code (and probably tests) to
+replace it with the new `LedgerEvent` date from the node.
+
+Similarly, both the `cardano-wallet` and `db-sync` already have an era independent block type
+(for `db-sync` this is Shelley/Allegra/Mary/Alonzo only) with conversion functions to convert
+from the era specific block/tx types to the generic block/tx types. Having `ledger-specs`
+provide this generic block type, reduces the amount of code `cardano-wallet` and `db-sync` needs
+to maintain.
+
+### Potential Problems Implementing the New API
+
+On startup of `db-sync` it needs to negotiate with the `node` on the most recent valid block.
+In the worst case, `db-sync` will need to sync from genesis or sync from some point for which
+the `node` does not have the required ledger state. This means that `node` will need to
+reconstruct the required ledger state.
+
+
+### Annotated Data Structures
+
+The idea here is that a single block type captures the super set of Byron, Shelley, Allegra,
+Mary, Alonzo and and future era. Fields that are not valid for all eras can be encoded as
+`Maybe`s or lists, so that fields that are not valid would be `Nothing` or `[]`.
+
+There are two main data structures required, `AnnotatedBlock` and `AnnotatedTx` which can be
+sketched out as follows:
+
+```
+data AnnotatedBlock = AnnotatedBlock
+  { abEra :: !Era
+  , abEpochNo :: !EpochNo
+  , abSlotNo :: !SlotNo
+  , abEpockSlot :: !Word   	    -- The slot within the epoch (starts at 0 for first slot of each epoch
+  , abTimeStamp :: !UTCTime     -- The slot number converted to UTCTime
+  , abEpochSize :: !EpochSize   -- Number of slots in current epoch
+  -- All fields in the superset of all block types
+  , abTxs :: [AnnotatedTx]
+  }
+
+data AnnotatedTx = AnnotatedTx
+  { atInputSum :: !Coin         -- Sum of the tx inputs
```

Comment:
Computing this at Node may be impossible. The reason is that this requires the current UTxO before a Tx is applied, which won't be available to the Node when we request for a block. This can only be available after db-sync has received the block and applies it, as a ledger event. 

Let me explain with an example:
Node tip is at Slot 10000 and db-sync requests block 100. The ledger state at slot 10000 can be used to produce the time metadata for block 100, because the time horizon for time metadata is pretty wide. However it has no way to resolve the inputs for block 100: the UTxO at slot 10000 is useless for this. The UTxO at slot 100 is also not enough, because it is already stale when a first transaction is applied. This can only be done in the general case as ledger events.

Reply:
My idea was that the ledger would calculate this. It gets a block from the applies it the the chain, and then returns a new ledger state, with an annotated block which includes annotated transactions.

Reply:
> Unless the inputs have already been spent, they should still be part of the UTxO set before the block is applied

The TxOut may not exist yet, if it is created by a previous Tx on the same block. So the UTxO state before the block application is not enough in the general case. The only place where the correct UTxO is available is somewhere inside the `applyBlock`, at the point where a Tx is applied. The only way to make this available is with ledger events, which is, in my mind at least, metadata of a block application.

Reply:
My idea is all predicated on there being a function like `applyBlock` but with a type signature:
```
applyAndAnnotateBlock :: LedgerState -> Block -> (LedgerState, [LedgerEvent])
```
where one of the constructors of `LedgerEvent` carries an `AnnotatedBlock` containing `AnnotatedTx`s.

Reply:
Yes this is exactly what we need eventually. 70% of db-sync code could be removed by a function like this. Unfortunately `applyblock` is a black box for `db-sync` which only returns the `LedgerState` currently.